### PR TITLE
samples: boards: nordic: coresight_stm: Extend to support PPR and FLPR

### PR DESCRIPTION
**Pull request by nordic-krch**
Added stmesp node to flpr DT.
Optimize standalone STM logging on PPR and FLPR (by not appending logging strings to the message).
Added support for FLPR and PPR in the sample.

### CODE REVIEW

**Comment by nordic-krch** [comment]
fyi @nordic-segl

**Review by nordic-piks — APPROVED** [review]

File: samples/boards/nordic/coresight_stm/remote/prj.conf
Comment (nordic-segl) [inline samples/boards/nordic/coresight_stm/remote/prj.conf:1]: Is this needed? This will disable logs from radio core on the second serial port when
sample.boards.nrf.coresight_stm.local_uart
is run.

File: samples/boards/nordic/coresight_stm/remote/boards/nrf54h20dk_nrf54h20_cpurad.overlay
Comment (nordic-segl) [inline samples/boards/nordic/coresight_stm/remote/boards/nrf54h20dk_nrf54h20_cpurad.overlay:1]: Is this needed? This will disable logs from radio core on the second serial port when
sample.boards.nrf.coresight_stm.local_uart
is run.
coresight_stm/remote will be build on Radio core. Let's keep it as it is.
coresight_stm/ppr and coresight_stm/flpr shall disable uart135 and disable KConfigs related to serial port console.

File: samples/boards/nordic/coresight_stm/ppr/boards/nrf54h20dk_nrf54h20_cpurad.overlay
Comment (nordic-segl) [inline samples/boards/nordic/coresight_stm/ppr/boards/nrf54h20dk_nrf54h20_cpurad.overlay:1]: Shouldn't his file be for cpuppr?

File: samples/boards/nordic/coresight_stm/boards/nrf54h20dk_nrf54h20_cpuapp.overlay
Comment (nordic-segl) [inline samples/boards/nordic/coresight_stm/boards/nrf54h20dk_nrf54h20_cpuapp.overlay:1]: not needed? Cpuapp doesn't enable uart135 by default.

File: samples/boards/nordic/coresight_stm/remote/prj.conf
Comment (nordic-krch) [reply samples/boards/nordic/coresight_stm/remote/prj.conf:1]: you are right, that will break `local_uart` configuration. However, I noticed an issue with ordering of applying overlays. Sample specific overlays are applied first and then snippets come. As a result I cannot disable uart135 in the application which is using `nordic-ppr` snippet which enables it.

File: samples/boards/nordic/coresight_stm/boards/nrf54h20dk_nrf54h20_cpuapp.overlay
Comment (nordic-krch) [reply samples/boards/nordic/coresight_stm/boards/nrf54h20dk_nrf54h20_cpuapp.overlay:1]: `nordic-ppr` snippet enables (reserves) it.

File: samples/boards/nordic/coresight_stm/remote/prj.conf
Comment (nordic-krch) [reply samples/boards/nordic/coresight_stm/remote/prj.conf:1]: fixed. The solution for ordering is to not use snippets at all.

File: samples/boards/nordic/coresight_stm/remote/boards/nrf54h20dk_nrf54h20_cpurad.overlay
Comment (nordic-krch) [reply samples/boards/nordic/coresight_stm/remote/boards/nrf54h20dk_nrf54h20_cpurad.overlay:1]: removed.

File: samples/boards/nordic/coresight_stm/ppr/boards/nrf54h20dk_nrf54h20_cpurad.overlay
Comment (nordic-krch) [reply samples/boards/nordic/coresight_stm/ppr/boards/nrf54h20dk_nrf54h20_cpurad.overlay:1]: removed

File: samples/boards/nordic/coresight_stm/boards/nrf54h20dk_nrf54h20_cpuapp.overlay
Comment (nordic-krch) [reply samples/boards/nordic/coresight_stm/boards/nrf54h20dk_nrf54h20_cpuapp.overlay:1]: not using that snippet so this entry can also be removed.

File: dts/riscv/nordic/nrf54h20_cpuflpr.dtsi [hunk 1]
```diff
@@ -27,6 +27,11 @@ cpusys_vevif: &cpusys_vevif_tx {};
 		compatible = "simple-bus";
 		interrupt-parent = <&cpuflpr_clic>;
 		ranges;
+
+		stmesp: memory@ae000000 {
+			compatible = "arm,stmesp";
+			reg = <0xad000000 0x1000000>;
+		};
```
Comment (gmarull) [inline dts/riscv/nordic/nrf54h20_cpuflpr.dtsi:34]: is this exclusive to cpuflpr? can't it be part of the global file?

File: samples/boards/nordic/coresight_stm/boards/nrf54h20dk_nrf54h20_cpuapp.overlay [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+	status = "okay";
+};
+
+&cpuppr_ram3x_region {
+	status = "okay";
+};
+
+&cpuflpr_vpr {
+	status = "okay";
+};
```
Comment (gmarull) [inline samples/boards/nordic/coresight_stm/boards/nrf54h20dk_nrf54h20_cpuapp.overlay:16]: why aren't nordic-flpr+ppr snippets used?

File: samples/boards/nordic/coresight_stm/Kconfig.sysbuild [hunk 3]
```diff
@@ -4,6 +4,22 @@
...
+
+config PPR_CORE_BOARD
+	string
+	default "nrf54h20dk/nrf54h20/cpuppr" if $(BOARD) = "nrf54h20dk"
+	depends on CORESIGHT_STM_USE_PPR
+
+config FLPR_CORE_BOARD
+	string
+	default "nrf54h20dk/nrf54h20/cpuflpr" if $(BOARD) = "nrf54h20dk"
+	depends on CORESIGHT_STM_USE_FLPR
```
Comment (gmarull) [inline samples/boards/nordic/coresight_stm/Kconfig.sysbuild:25]: loops a bit hacky, why is this needed? samples like sysbuild/hello_world stopped hardcoding board names in Kconfig.sysbuild, please check. You can use sample.yaml to specify supported builds.

Comment (nordic-segl) [reply samples/boards/nordic/coresight_stm/Kconfig.sysbuild:25]: I think this sample is a bit more complicated than sysbuild/hello_world. Besides app core there will be some number of "remotes".
On nrf54l20 we would like to heave all possible remotes - Radio, PPR, FLPR.
However, number of remotes can be different for other chips.
I believe, it would be good to have possibility to enable different number of remotes (APP+Radio or APP+Radio+PPR or APP+PPR or etc.) and select different target for each remote.

File: samples/boards/nordic/coresight_stm/boards/nrf54h20dk_nrf54h20_cpuapp.overlay (hunk 2 above)
Comment (nordic-segl) [reply samples/boards/nordic/coresight_stm/boards/nrf54h20dk_nrf54h20_cpuapp.overlay:16]: First version was using snippets. We encountered side effects resulting from snippet enabling uart nodes.

For example, nordic-ppr "reserves" uart135
https://github.com/zephyrproject-rtos/zephyr/blob/main/snippets/nordic-ppr/boards/nrf54h20dk_nrf54h20_cpuapp.overlay#L10
While uart135 is a default uart instance for cpurad. As a result we had conflict in UICRs and Secure Domain FW wasn't booting cores.

Comment (gmarull) [reply samples/boards/nordic/coresight_stm/boards/nrf54h20dk_nrf54h20_cpuapp.overlay:16]: should minimal ppr/flpr snippets be present for scenarios where UART is not needed? Do we envision other scenarios where this will be needed?

File: samples/boards/nordic/coresight_stm/Kconfig.sysbuild (hunk 3 above)
Comment (nordic-segl) [reply samples/boards/nordic/coresight_stm/Kconfig.sysbuild:25]: See sample.yaml - there are three configurations. When run on nrf54h20:

1. sample.boards.nrf.coresight_stm.dict will run on APP + Radio + PPR + FLPR
2. sample.boards.nrf.coresight_stm will run on APP + Radio + PPR + FLPR
3. sample.boards.nrf.coresight_stm.local_uart will run on APP + Radio

File: samples/boards/nordic/coresight_stm/boards/nrf54h20dk_nrf54h20_cpuapp.overlay (hunk 2 above)
Comment (nordic-segl) [reply samples/boards/nordic/coresight_stm/boards/nrf54h20dk_nrf54h20_cpuapp.overlay:16]: Whoever tries to use both Radio core and PPR, will face this problem.
Radio core uses uart135 as a default console:
https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/nordic/nrf54h20dk/nrf54h20dk_nrf54h20_cpurad.dts#L22
Snippet 'nordic-ppr' assigns uart135 to PPR:
https://github.com/zephyrproject-rtos/zephyr/blob/main/snippets/nordic-ppr/boards/nrf54h20dk_nrf54h20_cpuapp.overlay#L11
Also, @nordic-krch found that snippet is applied as a last so it's impossible to override uart135 state in a board overlay.

File: samples/boards/nordic/coresight_stm/Kconfig.sysbuild (hunk 3 above)
Comment (nordic-krch) [reply samples/boards/nordic/coresight_stm/Kconfig.sysbuild]: Reworked it to use single dir (`remote`) for rad,ppr and flpr. Removed hardcoded boards from `Kconfig.sysbuild`. @gmarull can you take a look?

File: samples/boards/nordic/coresight_stm/boards/nrf54h20dk_nrf54h20_cpuapp.overlay (hunk 2 above)
Comment (nordic-krch) [reply samples/boards/nordic/coresight_stm/boards/nrf54h20dk_nrf54h20_cpuapp.overlay:16]: nordic-ppr and nordic-flpr are simple snippets with 1 or 2 entries in the cpuapp overlay so i just moved them to here. When STM becomes a standard to be used for console output in nrf54h then maybe uart135 and uart120 nodes may be removed from those snippets. I'm affraid that more snippets will just cause trouble with too many options for the user.

File: dts/riscv/nordic/nrf54h20_cpuflpr.dtsi (hunk 1 above)
Comment (nordic-krch) [reply dts/riscv/nordic/nrf54h20_cpuflpr.dtsi:34]: it's a register set dedicated for flpr. Used same approach as in ppr case: https://github.com/zephyrproject-rtos/zephyr/blob/57a8a7c3500a6444448276103e890e5efbcd3a90/dts/riscv/nordic/nrf54h20_cpuppr.dtsi#L31

Comment (gmarull) [reply dts/riscv/nordic/nrf54h20_cpuflpr.dtsi:34]: ok, btw, 

```
stmesp: memory@ae000000` but `reg = <0xad000000 0x1000000>;
                ^~~                     ^~~
```

Comment (nordic-krch) [reply dts/riscv/nordic/nrf54h20_cpuflpr.dtsi]: fixed.

**Review by gmarull — APPROVED** [review]

**Review by nordic-segl — APPROVED** [review]

**Review by masz-nordic — APPROVED** [review]
